### PR TITLE
assignment3

### PR DESCRIPTION
**Pull request by haikyuu**
*  lib/index.js now exports init function
  * Add travis.yml
  * introduce start.js as new entry point
  * introduce lab & code for testing
  * 100% test coverage

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -7,22 +7,24 @@ var Version = require('./version');
...
-internals.init = function () {
+exports.init = internals.init = function (port, callback) {
 
     var server = new Hapi.Server();
-    server.connection({ port: 8000 });
+    server.connection({ port: port });
     server.register(Version, function (err) {
 
         Hoek.assert(!err, err);
+        internals.errors.push(err);
```
Comment (AdriVanHoudt) [inline lib/index.js]: you will never push something here since when there is an error `Hoek.assert()` will throw

File: lib/index.js [hunk 2]
```diff
@@ -7,22 +7,24 @@ var Version = require('./version');
...
         server.start(function (err) {
 
             Hoek.assert(!err, err);
-            console.log('Server started at: ' + server.info.uri);
+            internals.errors.push(err);
+            callback(internals.errors, server);
```
Comment (AdriVanHoudt) [inline lib/index.js]: use `return callback()` for readability

File: package.json [hunk 3]
```diff
@@ -21,8 +21,12 @@
...
     "hoek": "2.x.x"
   },
+  "devDependencies": {
+    "code": "^1.4.0",
+    "lab": "^5.5.1"
+  },
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1",
-    "start": "node lib/index.js"
+    "test": "make test-cov",
```
Comment (AdriVanHoudt) [inline package.json]: As I've stated before, I prefer Makefile pointing to npm since this won't work on windows

File: test/index.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+
+describe('server initiaion', function () {
+
+    it('initiate a server without errors', function (done) {
+		Init(8000, function (errors, server) {
+
+			expect(errors).to.only.include(undefined);
+			server.stop();
+			done();
```
Comment (AdriVanHoudt) [inline test/index.js]: do `server.stop(done)` so you are sure to only stop the test when the server is fully stopped

**Comment by haikyuu** [comment]
Thanks @AdriVanHoudt  .
I borrowed your test which alters the register function of the plugin, nice trick (i totally forgot that the required modules were cached).

**Comment by haikyuu** [comment]
Wouldn't it be better if we required all the plugins in the lib folder and registered them **once**.
We could use [require-dir] (https://github.com/aseemk/requireDir) or something like
```javascript
Fs.readdirSync(__dirname).filter(function(file) {

        return ((file !== 'index.js') && (file.indexOf('.js') !== -1);
    }).forEach(function(file) {
        internals.plugins.push(require(Path.join(__dirname, file)));
    });
```

**Comment by AdriVanHoudt** [comment]
@haikyuu np and we register them once right now...

**Comment by haikyuu** [comment]
@AdriVanHoudt  i meant that we shouldn't have to update the index.js file every time we add a plugin.

**Comment by AdriVanHoudt** [comment]
@haikyuu seems logical to me, since every plugin has specific options, we can however in the future use https://github.com/hapijs/glue

File: test/index.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+		var port = 8002;
+		Init(port, function (err, server) {
+
+			expect(err, 'err').not.to.exist();
+			expect(server.info.port, 'port').to.equal(port);
+			server.stop(done);
+		});
+
+    });
```
Comment (hueniverse) [inline test/index.js]: Need to work on style guide rules. New lines and spaces are wrong.

**Pull request closed without merging**